pypsa/networkclustering.py

```diff
@@ -513,6 +514,85 @@ def rectangular_grid_clustering(network, divisions):
     busmap = busmap_by_rectangular_grid(network.buses, divisions)
     return get_clustering_from_busmap(network, busmap)
 
+################
+# Hierarchical Clustering
+def busmap_by_hac(network, n_clusters, buses_i=None, branch_components=["Line", "Link"], feature=None):
+    """
+    Create a busmap accroding to Hierarchical Agglomerative Clustering.
+
+    Parameters
+    ----------
+    network : pypsa.Network
+    n_clusters : int
+        Final number of clusters desired.
+    buses_i: None|pandas.Index
+        Subset of buses to cluster. If None, all buses are considered.
+    branch_components: List
+        Subset of all branch_components in the network. Defaults to ["Line", "Link"].
+    feature: None | pandas.DataFrame
+        Feature to be considered for the clustering.
+        The DataFrame must be indexed with buses_i.
+        If None, all buses have the same similarity.
+
+    Returns
+    -------
+    busmap : pandas.Series
+        Mapping of network.buses to k-means clusters (indexed by
+        non-negative integers).
+    """
+
+    if find_spec('sklearn') is None:
+        raise ModuleNotFoundError("Optional dependency 'sklearn' not found."
+            "Install via 'conda install -c conda-forge scikit-learn' "
+            "or 'pip install scikit-learn'")
+
+    from sklearn.cluster import AgglomerativeClustering as HAC
+
+    if buses_i is None:
+        buses_i = network.buses.index
+
+    if feature is None:
+        logger.warning(
+            "No feature is specified for Hierarchical Clustering. "
+            "Falling back to default, where all buses have equal similarity. "
+            "You can specify a feature as pandas.DataFrame indexed with buses_i."
+        )
+
+        feature = pd.DataFrame(index=buses_i, columns=[""], data=0)
+
+    buses_x = network.buses.index.get_indexer(buses_i)
+
+    A = network.adjacency_matrix(branch_components=branch_components).todense()
+    A = sp.sparse.coo_matrix(A[buses_x].T[buses_x].T)
```

Comment:
Do you need to convert to dense before you reconvert to sparse? Should be more efficient without, but don't know if possible.

Reply:
unfortunately, I couldn't find a workaround... help wanted! :)

best would be, if pypsas adjacency_matrix had the `buses_i` as kwarg... maybe I can do something using `busorder`?

Reply:
```suggestion
    A = network.adjacency_matrix(branch_components=branch_components).tocsc()[buses_x, buses_x]
```
what about this? Compressed Sparse Column format allows indexing. And matrix multiplications should work.

Reply:
hmm, don't know a fix right now

I had thought about adding a kwarg to the adjacency_matrix function `as_pandas=False` that returns the adjacency matrix as pandas dataframe (similar for the other matrices of a network). Maybe that would help here?

Reply:
ahh, thanks. but does not work that way, we need 

```suggestion
    A = network.adjacency_matrix(branch_components=branch_components).tocsc()[buses_x]
```

Reply:
alternatively, pandas also supports sparse arrays which you can index by labels (as normal pandas data frames) 
```
A = pd.DataFrame.sparse.from_spmatrix(A, n.buses.index, n.buses.index).loc[buses_i, buses_i]
```

Reply:
> ahh, thanks. but does not work that way, we need

Ah, probably because of the tuple indexing (I updated the first suggestion)